### PR TITLE
Eradicate unused Imports and defined, but not used, functions, values

### PR DESCRIPTION
**Pull request by jrp2014**
I've left in Data.Semigroup and Data.Monoid, for older GHCs.  Agda should build a little faster.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ sudo: required
 
 branches:
   only:
+    - imports
```
Comment (andreasabel) [inline .travis.yml:20]: By pushing to a branch `issueXYZ`, you can forego the addition of your branch to `.travis.yml`.  See below
```
- /^issue.*/
```

File: src/full/Agda/Interaction/MakeCase.hs [hunk 2]
```diff
@@ -219,7 +219,7 @@ makeCase hole rng s = withInteractionId hole $ locallyTC eMakeCase (const True)
 
   -- Get function clause which contains the interaction point.
   InteractionPoint { ipMeta = mm, ipClause = ipCl} <- lookupInteractionPoint hole
-  let meta = fromMaybe __IMPOSSIBLE__ mm
+  let meta = fromMaybe __IMPOSSIBLE__ mm -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Interaction/MakeCase.hs]: Can be deleted.

File: src/full/Agda/Interaction/Imports.hs [hunk 3]
```diff
@@ -1119,7 +1119,7 @@ buildInterface source fileType topLevel pragmas = do
     reportSLn "import.iface" 5 "Building interface..."
     let m = topLevelModuleName topLevel
     scope'  <- getScope
-    let scope = set scopeCurrent m scope'
+    let scope = set scopeCurrent m scope' -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Interaction/Imports.hs]: Can be deleted.
Maybe also `scope' <- getScope` ?

File: src/full/Agda/Interaction/BasicOps.hs [hunk 4]
```diff
@@ -934,7 +934,7 @@ contextOfMeta ii norm = withInteractionId ii $ do
   withMetaInfo info $ do
     -- List of local variables.
     cxt <- getContext
-    let n         = length cxt
+    let n         = length cxt -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Interaction/BasicOps.hs]: Can be deleted.

File: src/full/Agda/Compiler/Treeless/Builtin.hs [hunk 5]
```diff
@@ -130,7 +130,7 @@ transform BuiltinKit{..} = tr
                 nPlusKAlt k b = TAGuard (tOp PGeq (TVar e) (tInt k)) $
                                 TLet (tOp PSub (TVar e) (tInt k)) b
 
-                str err = compactS err [Nothing]
+                str err = compactS err [Nothing] -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Compiler/Treeless/Builtin.hs]: Can be deleted.

File: src/full/Agda/Compiler/MAlonzo/Primitives.hs [hunk 6]
```diff
@@ -263,6 +263,7 @@ primBody s = maybe unimplemented (fromRight (hsVarUQ . HS.Ident) <$>) $
   ]
   where
   x |-> s = (x, Left <$> s)
+  -- TODO:: Defined but not used: ‘bin’
```
Comment (andreasabel) [inline src/full/Agda/Compiler/MAlonzo/Primitives.hs]: If you want, comment it out, but please don't delete, might be used in the future.

File: src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs [hunk 7]
```diff
@@ -186,7 +186,7 @@ recordConstructorType decls =
     makeBinding d = do
       let failure = typeError $ NotValidBeforeField d
           r       = getRange d
-          info    = ExprRange r
+          info    = ExprRange r -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs]: Can be deleted.

File: src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs [hunk 8]
```diff
@@ -2068,7 +2068,7 @@ instance ToAbstract C.Pragma [A.Pragma] where
       _       -> __IMPOSSIBLE__
   toAbstract (C.ForeignPragma _ rb s) = [] <$ addForeignCode (rangedThing rb) s
   toAbstract (C.CompilePragma _ rb x s) = do
-    let b = rangedThing rb
+    let b = rangedThing rb -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs]: Can be deleted.

File: src/full/Agda/Termination/TermCheck.hs [hunk 9]
```diff
@@ -127,9 +127,9 @@ termDecl' d = case d of
   where
     termDecls ds = concat <$> mapM termDecl' ds
 
-    unscopeDefs = concatMap unscopeDef
+    unscopeDefs = concatMap unscopeDef -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Termination/TermCheck.hs]: Can be deleted.

File: src/full/Agda/Termination/TermCheck.hs [hunk 10]
```diff
@@ -487,8 +487,8 @@ termType = return mempty
         extract dom `mappend` underAbstractionAbs dom absB (loop $! n + 1)
 
   -- create n variable patterns
-  mkPats n  = zipWith mkPat (downFrom n) <$> getContextNames
-  mkPat i x = notMasked $ VarP defaultPatternInfo $ DBPatVar (prettyShow x) i
+  mkPats n  = zipWith mkPat (downFrom n) <$> getContextNames -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Termination/TermCheck.hs]: Please keep.

File: src/full/Agda/TypeChecking/Conversion.hs [hunk 11]
```diff
@@ -1653,8 +1655,8 @@ equalSort s1 s2 = do
                   | otherwise -> postpone
            | otherwise -> postpone
 
-      set0 = mkType 0
-      prop0 = mkProp 0
+      set0 = mkType 0  -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Conversion.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Conversion.hs [hunk 12]
```diff
@@ -1780,6 +1782,7 @@ equalSort s1 s2 = do
           -- Anything else: postpone
           _        -> synEq s0 (FunSort s1 s2)
 
+      -- TODO:: Defined but not used: isBottomSort
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Conversion.hs]: You can comment this out, but please do not delete.

File: src/full/Agda/TypeChecking/Coverage.hs [hunk 13]
```diff
@@ -656,7 +656,7 @@ createMissingHCompClause f n x old_sc (SClause tel ps _sigma' cps (Just t)) = se
       (working_tel,_deltaEx) = splitTelescopeAt (size gamma + 3 + size delta) tel
 
       -- Γ,φ,u,u0,(x:H)(δ : Δ) ⊢ rhoS : Γ(x:H)(δ : Δ)
-      rhoS = liftS (size hdelta) $ raiseS 3
+      rhoS = liftS (size hdelta) $ raiseS 3 -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Coverage.hs]: Can be commented out.

File: src/full/Agda/TypeChecking/EtaContract.hs [hunk 14]
```diff
@@ -41,7 +41,7 @@ binAppView t = case t of
   Dummy{}    -> __IMPOSSIBLE__
   where
     noApp = NoApp t
-    app f [] = noApp
+    app f [] = noApp -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/EtaContract.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Injectivity.hs [hunk 15]
```diff
@@ -303,8 +303,8 @@ useInjectivity dir ty blk neu = locallyTC eInjectivityDepth succ $ do
         pwords "useInjectivity on" ++
         [ prettyTCM blk, prettyTCM cmp, prettyTCM neu, prettyTCM ty]
       let canReduceToSelf = Map.member (ConsHead f) hdMap || Map.member UnknownHead hdMap
-          allUnique       = all isUnique hdMap
-          isUnique [_] = True
+          allUnique       = all isUnique hdMap -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Injectivity.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Pretty.hs [hunk 16]
```diff
@@ -415,15 +415,15 @@ instance PrettyTCM a => PrettyTCM (Pattern' a) where
         prettyTCM c <+> fsep (map (prettyTCM . namedArg) ps)
         where
         b = conPRecord i && patOrigin (conPInfo i) /= PatOCon
-        showRec :: MonadPretty m => m Doc
+        showRec :: MonadPretty m => m Doc -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Pretty.hs]: Please keep.

File: src/full/Agda/TypeChecking/Primitive.hs [hunk 17]
```diff
@@ -547,7 +547,7 @@ primForce :: TCM PrimitiveImpl
 primForce = do
   let varEl s a = El (varSort s) <$> a
       varT s a  = varEl s (varM a)
-      varS s    = pure $ sort $ varSort s
+      varS s    = pure $ sort $ varSort s -- TODO:: defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Primitive.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Primitive.hs [hunk 18]
```diff
@@ -557,7 +557,7 @@ primForceLemma :: TCM PrimitiveImpl
 primForceLemma = do
   let varEl s a = El (varSort s) <$> a
       varT s a  = varEl s (varM a)
-      varS s    = pure $ sort $ varSort s
+      varS s    = pure $ sort $ varSort s -- TODO:: defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Primitive.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/Data.hs [hunk 19]
```diff
@@ -164,7 +164,7 @@ checkDataDef i name uc (A.DataDefParams gpars ps) cs =
             -- Return the data definition
             return dataDef{ dataPathCons = catMaybes pathCons }
 
-        let s      = dataSort dataDef
+        let s      = dataSort dataDef -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/Data.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs [hunk 20]
```diff
@@ -625,8 +625,8 @@ isHom n x = do
 
 findFlexible :: Int -> FlexibleVars -> Maybe (FlexibleVar Nat)
 findFlexible i flex =
-  let flex'      = map flexVar flex
-      flexible i = i `elem` flex'
+  let flex'      = map flexVar flex -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs [hunk 21]
```diff
@@ -900,7 +900,7 @@ unifyStep s (Injectivity k a d pars ixs c) = do
   -- Get constructor telescope and target indices
   cdef  <- liftTCM (getConInfo c)
   let ctype  = defType cdef `piApply` pars
-      forced = defForced cdef
+      forced = defForced cdef -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs [hunk 22]
```diff
@@ -891,7 +891,7 @@ unifyStep s step@Solution{} = solutionStep RetryNormalised s step
 unifyStep s (Injectivity k a d pars ixs c) = do
   ifM (liftTCM $ consOfHIT $ conName c) (return $ DontKnow []) $ do
   withoutK <- liftTCM withoutKOption
-  let n = eqCount s
+  let n = eqCount s -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs [hunk 23]
```diff
@@ -1036,7 +1036,7 @@ unifyStep s EtaExpandVar{ expandVar = fi, expandVarRecordType = d , expandVarPar
   where
     i = flexVar fi
     m = varCount s
-    n = eqCount s
+    n = eqCount s -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs [hunk 24]
```diff
@@ -1102,8 +1102,8 @@ unifyStep s (SkipIrrelevantEquation k) = do
 unifyStep s (TypeConInjectivity k d us vs) = do
   dtype <- defType <$> liftTCM (getConstInfo d)
   TelV dtel _ <- liftTCM $ telView dtype
-  let n   = eqCount s
-      m   = size dtel
+  let n   = eqCount s -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/SizedTypes/Solve.hs [hunk 25]
```diff
@@ -126,7 +126,7 @@ solveSizeConstraints flag =  do
       [ text $ "Solving constraints (" ++ show flag ++ ")"
       ] ++ map prettyTCM cs0
   let -- Error for giving up
-      cannotSolve :: TCM a
+      cannotSolve :: TCM a -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/SizedTypes/Solve.hs]: Please keep.

File: src/full/Agda/TypeChecking/Unquote.hs [hunk 26]
```diff
@@ -683,7 +683,7 @@ evalTCM v = do
     tcDeclareDef :: Arg QName -> R.Type -> UnquoteM Term
     tcDeclareDef (Arg i x) a = inOriginalContext $ do
       setDirty
-      let r = getRelevance i
+      let r = getRelevance i -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Unquote.hs]: Can be deleted.

File: src/full/Agda/TypeChecking/Unquote.hs [hunk 27]
```diff
@@ -705,7 +705,7 @@ evalTCM v = do
       when (Lens.getSafeMode clo) $ liftTCM $ typeError . GenericDocError =<<
         "Cannot postulate '" <+> prettyTCM x <+> ":" <+> prettyR a <+> "' with safe flag"
       setDirty
-      let r = getRelevance i
+      let r = getRelevance i -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Unquote.hs]: Can be deleted.

File: src/full/Agda/Utils/Update.hs [hunk 28]
```diff
@@ -49,7 +51,7 @@ runChangeT :: Functor m => ChangeT m a -> m (a, Bool)
 runChangeT = fmap (mapSnd getAny) . runWriterT . fromChangeT
 
 -- | Run a 'ChangeT' computation, but ignore change flag.
-execChangeT :: Functor m => ChangeT m a -> m a
+execChangeT :: Functor m => ChangeT m a -> m a -- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Utils/Update.hs]: Please keep, this is a library function.

File: src/full/Agda/Utils/Update.hs [hunk 29]
```diff
@@ -79,6 +81,7 @@ type EndoFun a = a -> a
 type Change  a = ChangeT Identity a
 type Updater a = UpdaterT Identity a
 
+-- TODO:: Defined but not used
```
Comment (andreasabel) [inline src/full/Agda/Utils/Update.hs]: Keep, library function.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks!  Good work!

The `std-lib` submodule update needs to be removed from this PR.

I went through most of the `TODO`s:  "Can be deleted" things can (and should) be deleted.  In general unused `let` one-liners can be removed (esp. if undocumented).  Subroutines that are unused could be just temporarily unused, thus, in general please keep.

File: .travis.yml (hunk 1 above)
Comment (jrp2014) [reply .travis.yml:20]: OK.

File: src/full/Agda/Interaction/MakeCase.hs (hunk 2 above)
Comment (jrp2014) [reply src/full/Agda/Interaction/MakeCase.hs]: Done

File: src/full/Agda/Interaction/Imports.hs (hunk 3 above)
Comment (jrp2014) [reply src/full/Agda/Interaction/Imports.hs]: yep.

File: src/full/Agda/Compiler/Treeless/Builtin.hs (hunk 5 above)
Comment (jrp2014) [reply src/full/Agda/Compiler/Treeless/Builtin.hs]: done

File: src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs (hunk 7 above)
Comment (jrp2014) [reply src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs]: zapped

File: src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs (hunk 8 above)
Comment (jrp2014) [reply src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs]: gone

File: src/full/Agda/Termination/TermCheck.hs (hunk 10 above)
Comment (jrp2014) [reply src/full/Agda/Termination/TermCheck.hs]: OK

File: src/full/Agda/TypeChecking/EtaContract.hs (hunk 14 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/EtaContract.hs]: gone

File: src/full/Agda/TypeChecking/Injectivity.hs (hunk 15 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Injectivity.hs]: Both gone

File: src/full/Agda/TypeChecking/Pretty.hs (hunk 16 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Pretty.hs]: Fine

File: src/full/Agda/TypeChecking/Primitive.hs (hunk 18 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Primitive.hs]: Zapped in both places....

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs (hunk 20 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: good

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs (hunk 21 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: history

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs (hunk 22 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: OK

File: src/full/Agda/TypeChecking/Rules/LHS/Unify.hs (hunk 24 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Rules/LHS/Unify.hs]: both gone

File: src/full/Agda/TypeChecking/SizedTypes/Solve.hs (hunk 25 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/SizedTypes/Solve.hs]: retained

File: src/full/Agda/TypeChecking/Unquote.hs (hunk 27 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Unquote.hs]: yup

File: src/full/Agda/Utils/Update.hs (hunk 28 above)
Comment (jrp2014) [reply src/full/Agda/Utils/Update.hs]: kept

File: src/full/Agda/Utils/Update.hs (hunk 29 above)
Comment (jrp2014) [reply src/full/Agda/Utils/Update.hs]: OK

**Comment by jrp2014** [comment]
Thanks for taking the trouble to do this, @andreasabel.  I have left the ones that you didn't flag specifically as I suspect that a few of them are for debugging.

Not sure why the last step of the continuous integration is failing.  I suspect that the golden results need to be updated (and ieee754 needs to be added to the test dependencies in `Agda.cabal`).  It's not v helpful, as I cannot tell whether an innocent change is making some subtle difference (which would be a pity, if the code base was so fragile) or whether HEAD is to expected to pass `make test`

File: src/full/Agda/Interaction/BasicOps.hs (hunk 4 above)
Comment (jrp2014) [reply src/full/Agda/Interaction/BasicOps.hs]: gone

File: src/full/Agda/Compiler/MAlonzo/Primitives.hs (hunk 6 above)
Comment (jrp2014) [reply src/full/Agda/Compiler/MAlonzo/Primitives.hs]: OK

File: src/full/Agda/Termination/TermCheck.hs (hunk 9 above)
Comment (jrp2014) [reply src/full/Agda/Termination/TermCheck.hs]: gone

File: src/full/Agda/TypeChecking/Conversion.hs (hunk 11 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Conversion.hs]: deleted

File: src/full/Agda/TypeChecking/Coverage.hs (hunk 13 above)
Comment (jrp2014) [reply src/full/Agda/TypeChecking/Coverage.hs]: done

**Comment by jrp2014** [comment]
I think this is ready.  The test failures are already there in the current clean master, which is a pity.

**Comment by jespercockx** [comment]
I think something went wrong when you merged in the current master into your local branch: now a bunch of other commits unrelated to this PR are also included and the PR cannot be merged automatically. Could you please try to fix the problem (e.g. by creating a fresh copy of the master branch and cherry-picking your commits on top of that one)?

**Comment by jrp2014** [comment]
Looks like I messed up my rebase.  Sorry.  I'll put together a patch.

**Comment by jrp2014** [comment]
The following file
[imports.txt](https://github.com/agda/agda/files/4258115/imports.txt)
is just a patch file that can be applied on top of master to comment out unused imports, delete some unused functions, and mark others with a TODO::.  I suggest that we keep the imports commented out for a while and then just delete them.  Similarly, if some of the unused functions are not reclaimed (eg, because they are not even useful for debugging) we can zap them.

**Comment by gallais** [comment]
> I suggest that we keep the imports commented out for a while and then just delete them.

I'd just nuke them tbh.

**Comment by L-TChen** [comment]
Yeah, just remove them. They will probably grow again in a year ..

**Comment by jrp2014** [comment]
The imports and crufty-looking underfineds are now zapped.

The git diff patch against master is here:
[imports.txt](https://github.com/agda/agda/files/4264674/imports.txt)

PS: The some tests fail because the line numbers are out.

**Comment by jrp2014** [comment]
If people are content with this, perhaps someone could apply the patch at https://github.com/agda/agda/pull/4468#issuecomment-592237864 and update the golden samples for some of the tests, because some line numbers have changed
```
< (agda2-info-action "*Error*" "We reached a program point we did not want to reach. Location of the error: src/full/Agda/TypeChecking/Unquote.hs:651 " nil)
---
> (agda2-info-action "*Error*" "We reached a program point we did not want to reach. Location of the error: src/full/Agda/TypeChecking/Unquote.hs:646 " nil)
```

**Comment by jrp2014** [comment]
I think that this is about there.  Some tests fail because line numbers have changed as a result of deleting stuff.  What's the best way to converge to resolution?

**Comment by jespercockx** [comment]
You can run `make test` locally, accept the changes, make a commit, and add it to the PR.

**Comment by jrp2014** [comment]
@andreasabel I think that this good to go...

**Review by andreasabel — APPROVED** [review]
Thanks, looks good!

**Pull request merged**